File: sample-kanban-node/app/services/board-services.js
/**
 * Services for kanban
 *
 * @param {{db: import('knex') }} options payload with query builder
 *
 * @returns configured services
 */
export const boardServices = ({db}) => {
	// Knex is a query builder, it's up to us to handle all 1+N selects
	async function loadTaskDeps(task) {
		task.status = await db('kanban.status').where({id: task.status_id}).first();
		task.people = await db('kanban.person')
			.whereIn('id', db('kanban.task_person')
				.select('person_id').where({task_id: task.id}));
		task.messages = await db('kanban.message').where({task_id: task.id});
		for (const message of task.messages) {
			message.person = await db('kanban.person').where({id: message.person_id}).first();
			message.task = task;
		}
	}

	return {
		async listUsers(q = '') {
			return db('kanban.person')
				.whereILike('name', `%${q}%`)
				.orderBy('name');
		},
		async findPerson(id) {
			return db('kanban.person').where({id}).first();
		},
		async listStatuses() {
			return db('kanban.status').orderBy('id');
		},
		async findStatus(id) {
			return db('kanban.status').where({id}).first();
		},
		async findStatusByTaskId(id) {
			return db('kanban.status')
				.where('id',
					db('kanban.task')
						.select('status_id')
						.where({id}).first())
				.first();
		},
		async listTasks(q = '') {
			const tasks = await db('kanban.task')
				.whereILike('description', `%${q}%`)
				.orderBy('id');
			for (const task of tasks) {
				await loadTaskDeps(task);
			}

			return tasks;
		},
		async findTask(id) {
			const task = await db('kanban.task').where({id}).first();
			await loadTaskDeps(task);
			return task;
		},
		async addTask({description, statusId}) {
			return db('kanban.task')
				.insert({description, status_id: statusId}).returning('*');
		},
		async updateTask({description, statusId, id}) {
			return db('kanban.task')
				.update({description, status_id: statusId})
				.where({id}).returning('*');
		},
		async deleteTask(id) {
			return db('kanban.task')
				.where({id})
				.delete().returning('*');
		},
		async removePerson({id, personId}) {
			return db('kanban.task_person')
				.where({person_id: personId, task_id: id})
				.delete().returning('*');
		},
		async joinTask({taskId, personId}) {
			return db('kanban.task_person')
				.insert({task_id: taskId, person_id: personId})
				.returning('*');
		},
		async addComment({taskId, personId, content}) {
			return db('kanban.message')
				.insert({task_id: taskId, person_id: personId, content})
				.returning('*');
		},
	};
};

File: sample-kanban-go/app/services/board-services.go
package services

import (
	"errors"
	"fmt"
	"github.com/doug-martin/goqu/v9"
	"github.com/sombriks/sample-testcontainers/sample-kanban-go/app/models"
)

type BoardService struct {
	db *goqu.Database
}

// NewBoardService - provision service-specific code
func NewBoardService(db *goqu.Database) (*BoardService, error) {
	service := &BoardService{
		db: db,
	}
	return service, nil
}

func (s *BoardService) hydrateMessage(message *models.Message) {
	message.Person, _ = s.FindPerson(message.PersonId)
}

func (s *BoardService) ListMessagesByTaskId(id int64) (*[]models.Message, error) {
	var messages []models.Message
	var err = s.db.From("kanban.message").
		Where(goqu.C("task_id").Eq(id)).
		ScanStructs(&messages)
	for i := range messages {
		s.hydrateMessage(&messages[i])
	}
	return &messages, err
}

func (s *BoardService) ListPeople(q string) (*[]models.Person, error) {
	var people []models.Person
	var err = s.db.From("kanban.person").
		Where(goqu.Ex{"name": goqu.Op{"ilike": fmt.Sprint("%", q, "%")}}).
		ScanStructs(&people)
	return &people, err
}

func (s *BoardService) ListPeopleByTaskId(id int64) (*[]models.Person, error) {
	var people []models.Person
	var err = s.db.From("kanban.person").
		Where(goqu.C("id").
			In(s.db.From("kanban.task_person").
				Select("person_id").
				Where(goqu.C("task_id").Eq(id)))).
		ScanStructs(&people)
	return &people, err
}

func (s *BoardService) FindPerson(id int64) (*models.Person, error) {
	var person models.Person
	var ok, err = s.db.From("kanban.person").
		Where(goqu.C("id").Eq(id)).
		ScanStruct(&person)
	if !ok {
		return nil, errors.New(fmt.Sprint("Person #", id, " not found"))
	}
	return &person, err
}

func (s *BoardService) ListStatus() (*[]models.Status, error) {
	var statuses []models.Status
	err := s.db.From("kanban.status").
		Order(goqu.C("id").Asc()).
		ScanStructs(&statuses)
	return &statuses, err
}

func (s *BoardService) FindStatus(id int64) (*models.Status, error) {
	var status models.Status
	ok, err := s.db.From("kanban.status").
		Where(goqu.C("id").Eq(id)).
		ScanStruct(&status)
	if !ok {
		return nil, errors.New(fmt.Sprint("Status #", id, " not found"))
	}
	return &status, err
}
func (s *BoardService) FindStatusByTaskId(id int64) (*models.Status, error) {
	var status models.Status
	ok, err := s.db.From("kanban.status").
		Where(goqu.C("id").Eq(s.db.From("kanban.task").
			Select(goqu.C("status_id")).
			Where(goqu.C("id").Eq(id)))).
		ScanStruct(&status)
	if !ok {
		return nil, errors.New(fmt.Sprint("status for task #", id, " not found"))
	}
	return &status, err
}

func (s *BoardService) ListTasks(q string) (*[]models.Task, error) {
	var tasks []models.Task
	var err = s.db.From("kanban.task").
		Where(goqu.C("description").ILike(fmt.Sprint("%", q, "%"))).
		Order(goqu.C("id").Asc()).
		ScanStructs(&tasks)
	for i := range tasks { // https://gobyexample.com/range (beware, values, not references!)
		s.hydrateTask(&tasks[i])
	}
	return &tasks, err
}

func (s *BoardService) hydrateTask(task *models.Task) {
	task.Status, _ = s.FindStatus(task.StatusId)
	task.People, _ = s.ListPeopleByTaskId(task.Id)
	task.Messages, _ = s.ListMessagesByTaskId(task.Id)

}

func (s *BoardService) FindTask(id int64) (*models.Task, error) {
	var task models.Task
	var ok, err = s.db.From("kanban.task").
		Where(goqu.C("id").Eq(id)).
		ScanStruct(&task)
	if !ok {
		return nil, errors.New(fmt.Sprint("Task #", id, " not found"))
	}
	s.hydrateTask(&task)
	return &task, err
}

func (s *BoardService) InsertTask(task *models.Task) (*models.Task, error) {
	ok, err := s.db.Insert("kanban.task").
		Rows(task).
		Returning("*").
		Executor().
		ScanStruct(task)
	if !ok {
		return nil, errors.New("failed to insert task")
	}
	return task, err
}

func (s *BoardService) UpdateTask(id int64, task *models.Task) (*models.Task, error) {
	task.Id = id
	_, err := s.db.Update("kanban.task").
		Set(*task).Where(goqu.C("id").Eq(id)).Executor().Exec()
	return task, err
}

func (s *BoardService) DeleteTask(id int64) (int64, error) {
	result, err := s.db.Delete("kanban.task").
		Where(goqu.C("id").Eq(id)).
		Executor().
		Exec()
	affected, err := result.RowsAffected()
	return affected, err
}

func (s *BoardService) RemovePerson(taskId int64, personId int64) error {
	_, err := s.db.Delete("kanban.task_person").Where(goqu.Ex{
		"person_id": personId, "task_id": taskId,
	}).Executor().Exec()
	return err
}

func (s *BoardService) JoinTask(taskId int64, personId int64) error {
	_, err := s.db.Insert("kanban.task_person").
		Cols("task_id", "person_id").
		Vals([]interface{}{taskId, personId}).
		Executor().Exec()
	return err
}

func (s *BoardService) AddComment(taskId int64, personId int64, content string) error {
	_, err := s.db.Insert("kanban.message").
		Rows(&models.Message{
			PersonId: personId,
			Content:  content,
			TaskId:   taskId,
		}).
		Executor().Exec()
	return err
}
